In [1]:
## Import library
import pandas as pd

In [2]:
## Read the file .csv and create the DataFrame
df = pd.read_csv('music_project_en.csv')

In [3]:
## Show the first 10 rows
print(df.head(10))

     userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5

In [4]:
## Show the general information included in the DataFrame
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63736 non-null  object
 2   artist    57512 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB


The information included in the DataFrame shows 7 columns with data type object: 

**Details:**
- `' userID'`: id of the user;
- `'Track'`: title of the song;
- `'artist'`: name of the artist;
- `'genre'`: genre of the song;
- `'City'`: city of the user;
- `'time'`: the exact time when the song was played;
- `'Day'`: day of the week.

In [5]:
## Show the name of the columns
print(df.columns)

Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')


Looking at the list of the columns name, we note some points that would be useful to improve.


In [6]:
## Cleaning the columns name: Convert to lower case, quit spaces and apply snake_case.
df.columns = df.columns.str.lower().str.strip().str.replace(' ', '_')
## Rename the column 'userid'
df.rename(columns={'userid': 'user_id'}, inplace=True)
## Print the columns to prove the changes made
print(df.columns)

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


In [7]:
## Calculate the number of missing values
print(df.isna().sum())

user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64


**Missing values:** 
The DataFrame includes missing values for the columns: track, artist and genre. In the next step we are going to fill these values 
with the word 'unknown'.

In [8]:
## Create a list with the columns that include missing values.
columns_to_replace= ['track','artist','genre']
## By using a loop fill the missing values with the word 'unknown'.
for column_to_replace in columns_to_replace:
    df.fillna('unknown', inplace= True)
## Calculate the missing values in order to prove the changes made.
print(df.isna().sum())

user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64


In [9]:
## Calculate the duplicated values that includes the DataFrame.
print(df.duplicated().sum())

3826


**Explicit duplicated values:** <br>
The DataFrame includes 3,826 duplicated values. In the next step we are going to delete those values.

In [10]:
## Delete duplicated values
df = df.drop_duplicates()
## Calculate the duplicated values to prove the changes made.
print(df.duplicated().sum())

0


**Implicit duplicated values in the column 'genre':** <br>
It is necesary to know if there are duplicated values in the column 'genre', that's why we are going to show the unique values for that column.

In [11]:
## Show the quantity of duplicated values included in the column 'genre'.
print(f"The column 'genre' includes {df['genre'].nunique()} duplicated values.")

The column 'genre' includes 269 duplicated values.


In [12]:
## Show the unique values for the column 'genre'.
print(sorted(df['genre'].unique()))

['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans', 'alternative', 'ambient', 'americana', 'animated', 'anime', 'arabesk', 'arabic', 'arena', 'argentinetango', 'art', 'audiobook', 'avantgarde', 'axé', 'baile', 'balkan', 'beats', 'bigroom', 'black', 'bluegrass', 'blues', 'bollywood', 'bossa', 'brazilian', 'breakbeat', 'breaks', 'broadway', 'cantautori', 'cantopop', 'canzone', 'caribbean', 'caucasian', 'celtic', 'chamber', 'children', 'chill', 'chinese', 'choral', 'christian', 'christmas', 'classical', 'classicmetal', 'club', 'colombian', 'comedy', 'conjazz', 'contemporary', 'country', 'cuban', 'dance', 'dancehall', 'dancepop', 'dark', 'death', 'deep', 'deutschrock', 'deutschspr', 'dirty', 'disco', 'dnb', 'documentary', 'downbeat', 'downtempo', 'drum', 'dub', 'dubstep', 'eastern', 'easy', 'electronic', 'electropop', 'emo', 'entehno', 'epicmetal', 'estrada', 'ethnic', 'eurofolk', 'european', 'experimental', 'extrememetal', 'fado', 'film', 'fitness', 'flamenco', 'folk', 'folklor

We are going to examine the list to identify **implicit duplicates** of the hip-hop genre, that are misspelled names or variants that refer to the same musical genre.

The duplicated values that we will find below:

* `hip`  
* `hop`  
* `hip-hop`  

Define a function called 'replace_wrong_values()' with the parameters below:

* `df`: The DataFrame to modify
* `column`: The name of the column used
* `wrong_values`: A list of the wrong values
* `correct_value`: The correct value to replace the wrong

In [13]:
## Function to replace the implicit duplicated values.
def replace_wrong_values(df,column,wrong_values,correct_value):
    for wrong_value in wrong_values:
        df[column]= df[column].replace(wrong_value,correct_value)
    return df

In [14]:
## By using the function delete the implicit duplicated values in the column 'genre'.
replace_wrong_values(df,'genre',['hip', 'hop', 'hip-hop'],'hiphop')

,user_id,track,artist,genre,city,time,day
0,FFB692EC,Kamigata To Boots,The Mass Missile,rock,Shelbyville,20:28:33,Wednesday
1,55204538,Delayed Because of Accident,Andreas Rönnberg,rock,Springfield,14:07:09,Friday
2,20EC38,Funiculì funiculà,Mario Lanza,pop,Shelbyville,20:58:07,Wednesday
3,A3DD03C9,Dragons in the Sunset,Fire + Ice,folk,Shelbyville,08:37:09,Monday
4,E2DC1FAE,Soul People,Space Echo,dance,Springfield,08:34:34,Monday
...,...,...,...,...,...,...,...
65074,729CBB09,My Name,McLean,rnb,Springfield,13:32:28,Wednesday
65075,D08D4A55,Maybe One Day (feat. Black Spade),Blu & Exile,hiphop,Shelbyville,10:00:00,Monday
65076,C5E3A0D5,Jalopiina,unknown,industrial,Springfield,20:09:26,Friday
65077,321D0506,Freight Train,Chas McDevitt,rock,Springfield,21:43:59,Friday


In [15]:
## Show the list of the unique values in the column 'genre' in order to confirm the changes made.
print(df['genre'].unique())

['rock' 'pop' 'folk' 'dance' 'rusrap' 'ruspop' 'world' 'electronic'
 'unknown' 'alternative' 'children' 'rnb' 'hiphop' 'jazz' 'postrock'
 'latin' 'classical' 'metal' 'reggae' 'triphop' 'blues' 'instrumental'
 'rusrock' 'dnb' 'türk' 'post' 'country' 'psychedelic' 'conjazz' 'indie'
 'posthardcore' 'local' 'avantgarde' 'punk' 'videogame' 'techno' 'house'
 'christmas' 'melodic' 'caucasian' 'reggaeton' 'soundtrack' 'singer' 'ska'
 'salsa' 'ambient' 'film' 'western' 'rap' 'beats' "hard'n'heavy"
 'progmetal' 'minimal' 'tropical' 'contemporary' 'new' 'soul' 'holiday'
 'german' 'jpop' 'spiritual' 'urban' 'gospel' 'nujazz' 'folkmetal'
 'trance' 'miscellaneous' 'anime' 'hardcore' 'progressive' 'korean'
 'numetal' 'vocal' 'estrada' 'tango' 'loungeelectronic' 'classicmetal'
 'dubstep' 'club' 'deep' 'southern' 'black' 'folkrock' 'fitness' 'french'
 'disco' 'religious' 'drum' 'extrememetal' 'türkçe' 'experimental' 'easy'
 'metalcore' 'modern' 'argentinetango' 'old' 'swing' 'breaks' 'eurofolk'
 'stone

In [16]:
## Show the quantity of unique values in the column 'genre' to confirm the changes made.
print(df['genre'].nunique())

266


In [31]:
## Group by the column 'city'
shelbyville = df[df['city'] == 'Shelbyville']
springfield = df[df['city'] == 'Springfield']
## Calculate the songs played for each city
print(f"The total of the songs played in Shelbyville is: {len(shelbyville)}")
print()
print(f"The total of the songs played in Springfield is: {len(springfield)}")

The total of the songs played in Shelbyville is: 18512

The total of the songs played in Springfield is: 42741


**Played songs for each city:** <br>
Accorging to the last countdown, we see that Springfield has a huge advantage about the songs played and the difference could be caused
because this place has allmost 6 times (113,000) more population than Shelbyville (20,000).

In [18]:
# Calculate the songs played for the each day: 'Monday' and 'Friday'
monday = df[df['day'] == 'Monday']
print(f"The quantity of songs played on monday is: {len(monday)}")
print()
friday = df[df['day'] == 'Friday']
print(f"The quantity of songs played on friday is: {len(friday)}")

The quantity of songs played on monday is: 21354

The quantity of songs played on friday is: 21840


In [29]:
## Calculate the total play songs during monday for each city.
monday_shell = df[(df['day'] == 'Monday') & (df['city'] == 'Shelbyville')]
print(f" The total of songs played on monday for Shelbyville is: {len(monday_shell)}")
print()
monday_spring = df[(df['day'] == 'Monday') & (df['city'] == 'Springfield')]
print(f" The total of songs played on monday for Springfield is: {len(monday_spring)}")

 The total of songs played on monday for Shelbyville is: 5614

 The total of songs played on monday for Springfield is: 15740


In [30]:
## Calculate the total play songs during friday for each city.
friday_shell = df[(df['day'] == 'Friday') & (df['city'] == 'Shelbyville')]
print(f" The total of songs played on friday for Shelbyville is: {len(friday_shell)}")
print()
friday_spring = df[(df['day'] == 'Friday') & (df['city'] == 'Springfield')]
print(f" The total of songs played on friday for Shelbyville is: {len(friday_spring)}")

 The total of songs played on friday for Shelbyville is: 5895

 The total of songs played on friday for Shelbyville is: 15945


**Played songs during monday and friday for each city:** <br>
On friday we can see a higher number of reproductions and it could be caused because the weekend is really close and the people get relax
by listening music. It is curious that we have a similar behaivor eventhough we group the results by city.


# Conclusion <a id='end'></a>

According to the information obtained, we can see that there is a greater number of plays for the city of Springfield, showing three times as many tracks played compared to Shelbyville.
Among the issues detected were the more than **3,000 explicit duplicates** that were removed, as well as the existence of implicit duplicates that were corrected to ensure accurate data counting. This allowed us to gain greater clarity in the information we wanted to analyze.After that, we determined that the variation from day to day for both cities is very small; it means, Shelbyville only has a difference of around 200 plays between Monday and Friday. Additionally, the same pattern is observed for the city of Springfield during those days.